Introduce SafeDbt to handle Dbt with free or memory_cleanse raii-style

Empact opened the pull request:
This provides additional exception-safety and case handling for the proper
freeing of the associated buffers.

Empact committed:
Drop unused setRange arg to BerkeleyBatch::ReadAtCursor

Empact committed:
Introduce SafeDbt to handle DB_DBT_MALLOC raii-style

This provides additional exception-safety and case handling for the proper
freeing of the associated buffers.

practicalswift commented on src/wallet/db.h:
```diff
@@ -166,6 +166,42 @@ class BerkeleyDatabase
+            return &m_dbt;
+        }
+    };
+}
```
```
2018-09-22 20:42:07 cpplint(pr=14268): src/wallet/db.h:204:  Anonymous namespace should be terminated with "// namespace"  [readability/namespace] [5]
```

251Labs commented on src/wallet/db.h:
```diff
@@ -166,6 +166,42 @@ class BerkeleyDatabase
+namespace {
+    /** RAII class that provides access to a DB_DBT_MALLOC Dbt */
+    class MallocedDbt {
+    private:
```
nit: You can drop the `private` access modifier. `m_dbt` is implicitly declared private, because private is the default class access level.

l2a5b1 reviewed (commented):
utACK 985892e

promag commented on src/wallet/db.h:
```diff
@@ -166,6 +166,42 @@ class BerkeleyDatabase
 
+namespace {
+    /** RAII class that provides access to a DB_DBT_MALLOC Dbt */
+    class MallocedDbt {
```
nit, IMO class name could be improved, as it also calls `memory_cleanse`. Maybe `SafeDbt` — I'm sure there is a better name 😛

promag reviewed (commented):
utACK 985892e.

nit, first commit kind of unrelated — could mention the dead code cleanup in the PR description.

achow101 commented:
utACK 985892e5aa75b76f71a354c8d835d7729048cccc

MeshCollider commented:
utACK https://github.com/bitcoin/bitcoin/pull/14268/commits/985892e5aa75b76f71a354c8d835d7729048cccc

Style-nit: the developer notes say "No indentation for `public`/`protected`/`private` or for `namespace`.", so you might want to change the MallocedDbt class to fit that.

First commit is unrelated as promag said, but a nice cleanup, the last use of `setRange` was in `ListAccountCreditDebit` which was removed when accounts were.

Empact commented:
Ok just pushed some significant changes:
* Now called `SafeDbt`, and applied in all cases where `memory_cleanse` was used with `Dbt`
* No longer in an anonymous namespace as that is a violation of [DCL59-CPP](https://wiki.sei.cmu.edu/confluence/display/cplusplus/DCL59-CPP.+Do+not+define+an+unnamed+namespace+in+a+header+file), instead nest it inside BerkeleyBatch, as all uses are therein.
* Implementation is now in `db.cpp`

Empact commented:
Appveyor failure looks unrelated.

ken2812221 commented:
utACK 1a9f9f7e5e2e73fb832f5b96ad7e9e57954f3f3c

practicalswift commented on src/wallet/db.cpp:
```diff
@@ -247,6 +247,45 @@ BerkeleyEnvironment::VerifyResult BerkeleyEnvironment::Verify(const std::string&
+    return m_dbt.get_size();
+}
+
+BerkeleyBatch::SafeDbt::operator Dbt*()
```
`const`? :-)

laanwj replied:
it's returning a mutable pointer to the internal structure, I don't think it'd make sense to make the method const (unless you mean it should return a `const*` pointer too, but that's likely not good enough for the BDB functions)

laanwj commented:
utACK 1a9f9f7e5e2e73fb832f5b96ad7e9e57954f3f3c

laanwj commented on src/wallet/db.cpp:
```diff
@@ -247,6 +247,45 @@ BerkeleyEnvironment::VerifyResult BerkeleyEnvironment::Verify(const std::string&
+{
+    if (m_dbt.get_data() != nullptr) {
+        // Clear memory, e.g. in case it was a private key
+        memory_cleanse(m_dbt.get_data(), m_dbt.get_size());
```
This changes behavior: in the case of `Write` it was also cleansing the key.
Not sure if this is necessary, I don't think the keys ever contain sensitive information, but it should be noted.

Empact committed:
Make SafeDbt DB_DBT_MALLOC on default initialization

If we're constructing the SafeDbt without provided data, it is always malloced,
so that is the case we expose.

Also run clang-format.

Empact commented:
Added another commit:
* Made `SafeDbt` constructors `explicit`
* Removed an unnecessary default value
* Ran `clang-format`
Separate for easier review, can squash.

Empact commented:
Appveyor failure looks unrelated: "ConnectionAbortedError: [WinError 10053] An established connection was aborted by the software in your host machine"
